# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: docshell/base/BrowsingContextGroup.cpp
# Commit: 36b0646e8649
# Full Hash: 36b0646e86498662e770756ad54b1fd648a82184
# Author: Andreas Farre <farre@mozilla.com>
# Date: 2020-04-07 21:44:02
# Regressor Bug: 1620594
# File Overlap Count: 1
# Description:
#   Bug 1620594 - Part 7: Remove TabGroup and SystemGroup. r=nika,bas
#   
#   TabGroup never really made any difference in which thread something go
#   dispatched to. This was the intended use, but development of TabGroups
#   with abstract main threads never made it that far. The good thing is
# ==============================================================================

diff -r a5ad3b179c92 -r 36b0646e8649 docshell/base/BrowsingContextGroup.cpp
--- a/docshell/base/BrowsingContextGroup.cpp	Tue Apr 07 15:17:14 2020 +0000
+++ b/docshell/base/BrowsingContextGroup.cpp	Tue Apr 07 15:17:47 2020 +0000
@@ -8,6 +8,7 @@
 #include "mozilla/dom/BrowsingContextBinding.h"
 #include "mozilla/dom/BindingUtils.h"
 #include "mozilla/dom/ContentParent.h"
+#include "mozilla/dom/DocGroup.h"
 #include "mozilla/StaticPrefs_dom.h"
 #include "mozilla/ThrottledEventQueue.h"
 #include "nsFocusManager.h"
@@ -21,6 +22,12 @@
   } else {
     ContentParent::HoldBrowsingContextGroup(this);
   }
+
+  mTimerEventQueue = ThrottledEventQueue::Create(
+      GetMainThreadSerialEventTarget(), "BrowsingContextGroup timer queue");
+
+  mWorkerEventQueue = ThrottledEventQueue::Create(
+      GetMainThreadSerialEventTarget(), "BrowsingContextGroup worker queue");
 }
 
 bool BrowsingContextGroup::Contains(BrowsingContext* aBrowsingContext) {
@@ -40,6 +47,7 @@
     // There are no browsing context still referencing this group. We can clear
     // all subscribers.
     UnsubscribeAllContentParents();
+
     if (XRE_IsContentProcess()) {
       ContentChild::GetSingleton()->ReleaseBrowsingContextGroup(this);
     } else {
@@ -219,27 +227,42 @@
 }
 
 void BrowsingContextGroup::GetDocGroups(nsTArray<DocGroup*>& aDocGroups) {
-  nsTHashtable<nsRefPtrHashKey<DocGroup>> docGroups;
-  for (auto& toplevel : mToplevels) {
-    toplevel->PreOrderWalk([&](BrowsingContext* aContext) {
-      if (nsDocShell* docShell = nsDocShell::Cast(aContext->GetDocShell())) {
-        if (!docShell->HasContentViewer()) {
-          return;
-        }
-        if (Document* document = docShell->GetDocument()) {
-          docGroups.PutEntry(document->GetDocGroup());
-        }
-      }
-    });
+  MOZ_ASSERT(NS_IsMainThread());
+  for (auto iter = mDocGroups.ConstIter(); !iter.Done(); iter.Next()) {
+    aDocGroups.AppendElement(iter.Data());
+  }
+}
+
+already_AddRefed<DocGroup> BrowsingContextGroup::AddDocument(
+    const nsACString& aKey, Document* aDocument) {
+  MOZ_ASSERT(NS_IsMainThread());
+
+  RefPtr<DocGroup>& docGroup = mDocGroups.GetOrInsert(aKey);
+  if (!docGroup) {
+    docGroup = DocGroup::Create(this, aKey);
   }
 
-  for (auto iter = docGroups.ConstIter(); !iter.Done(); iter.Next()) {
-    aDocGroups.AppendElement(iter.Get()->GetKey());
+  docGroup->AddDocument(aDocument);
+  return do_AddRef(docGroup);
+}
+
+void BrowsingContextGroup::RemoveDocument(const nsACString& aKey,
+                                          Document* aDocument) {
+  MOZ_ASSERT(NS_IsMainThread());
+  RefPtr<DocGroup> docGroup = aDocument->GetDocGroup();
+  // Removing the last document in DocGroup might decrement the
+  // DocGroup BrowsingContextGroup's refcount to 0.
+  RefPtr<BrowsingContextGroup> kungFuDeathGrip(this);
+  docGroup->RemoveDocument(aDocument);
+
+  if (docGroup->IsEmpty()) {
+    mDocGroups.Remove(aKey);
   }
 }
 
 NS_IMPL_CYCLE_COLLECTION_WRAPPERCACHE(BrowsingContextGroup, mContexts,
-                                      mToplevels, mSubscribers, mCachedContexts)
+                                      mToplevels, mSubscribers, mCachedContexts,
+                                      mTimerEventQueue, mWorkerEventQueue)
 
 NS_IMPL_CYCLE_COLLECTION_ROOT_NATIVE(BrowsingContextGroup, AddRef)
 NS_IMPL_CYCLE_COLLECTION_UNROOT_NATIVE(BrowsingContextGroup, Release)